### DISCUSSION TITLE
Does the runtime expose default values for properties?

### DISCUSSION THREAD

**Discussion by NSExceptional** [Azoy/Echo#53] [Q&A] [answered]
Would be a cool way for my JSON library to allow the user to supply default values, i.e.

```swift
struct Post {
    // Everyone
    var participants: [User]
    // Everyone except OP, computed later
    var commentors: [User] = []
}

**Comment by Azoy** [comment, marked as answer]
The runtime does not expose said information, however there are functions emitted for these default values. So in theory it is possible to get those values, but it would require a bit of work.

**Reply by NSExceptional** [reply]
You mean, the runtime generates functions to initialize each property?

I assume this would require some good ol' `dlsym()`

**Reply by Azoy** [reply]
The compiler emits default value functions yes. It would use some `dlsym` but would require understanding of the mangling to get it right.
```swift
struct Dog {
  var age: Int = 128
}

let RTLD_DEFAULT = UnsafeMutableRawPointer(bitPattern: -2)
let ptr = dlsym(RTLD_DEFAULT, "$s10SomeModule3DogV3ageSivpfi")
let fn = unsafeBitCast(ptr, as: (@convention(thin) () -> Int).self)
print(fn()) // 128
```

**Reply by NSExceptional** [reply]
Gotcha, okay. Does Echo provide any mechanisms to generate a mangled name like this? Or does your other private project? Or do you know where I can learn about how the names for these functions are constructed? (Or is it just the mangled version of `Module.Type.Property`?)

**Reply by Azoy** [reply]
I do have some work on a mangler, but I’m unsure if/when I’ll release it.

**Reply by NSExceptional** [reply]
No pressure, but it would be a shame if that code never saw the light of day!
